Stop audio output before fifo_stop is called

ragb opened the pull request:
This allows process_espeak_command to stop as soon as possible

Using emacspeak server with espeak-ng (still not published) found that audio was only being stoped after a command is processed (i.e. using pcaudiolib).
This fixes the issue.

rhdunn commented:
Is there a reason the lock/unlock on my_mutex in fifo.c was commented out?

If that code should be removed, please delete the lines instead of commenting them out. This is to make the resulting code cleaner and easier to read.

ragb commented:
My fault, that is not neede at all.
Removal is already committed.

rhdunn commented:
No problem. Thanks for the bug fix.

CMB commented:
I see a couple problems with this pull request.
First, we have audio_object_flush being called before fifo_stop,
so the audio could be being flushed while another thread is writing to
the audio device.
Second, I'm seeing hard lockups after less than a minute of use,
so there are new concurrency issues somewhere.

ragb commented:
I have no problems with this and coreaudio. Tested a bit with pulseaudio on linux and it worked fine, as far as I recall. However, if it locks up somewhere where it didn’t please revert.

ragb commented:
Meanwhile, concurrency problems may be caused by other pull request where I changed the code to not use sem_timedwait which doesn’t exist on OS X. I changed it to use condition variables.

rhdunn commented:
I have reverted this -- it is causing issues with `brltty`. This is likely due to the way that `brltty` is calling the espeak API (i.e. using it as a server on a thread).

CMB commented:
Well it's possible that pulse and coreaudio are threadsafe, I don't know.
I wonder if we could just add another mutex to the audio calls,
espeak_ng_audio_mutex.
Or maybe locking in pcaudiolib itself.

The other concurrency issues I mentioned were possibly caused by the
fact that the mutex wasn't being relocked after processing an espeak command,
in say_thread.  It wasn't locked for subsequent iterations of
the while(my_command_is_running) loop in say_thread.
say_thread is very tricky with the mutex.  I was starting to
restructure it a bit when the revert came through.
I'll keep working on a branch with your pull request;
it looks promising.

ragb commented:
At least coreaudio is suposed to be thread-safe (I tried to do that at least).

My idea on changing the order of the calls was to stop audio as soon as possible. I’m using emacspeak on OS X with this daily and it’s very very responsive stopping audio.
